### PR TITLE
Add derivative of `ArrayUtils.concat` that doesn't require explicit destination

### PR DESCRIPTION
**Pull request by JackPGreen**
`ArrayUtils.concat` requires a destination array as a parameter of the required size. To avoid callers calculating this, added a derivative that creates a new array internally and returns that.

This allows a simpler invocation where you simply want to combine two arrays.

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/6814

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java [hunk 1]
```diff
@@ -339,8 +339,7 @@ static <T extends CompoundPredicate> T flattenCompound(Predicate predicateLeft,
             Predicate[] left = getSubPredicatesIfClass(predicateLeft, klass);
             Predicate[] right = getSubPredicatesIfClass(predicateRight, klass);
 
-            predicates = new Predicate[left.length + right.length];
-            ArrayUtils.concat(left, right, predicates);
+            predicates = ArrayUtils.concat(left, right);
```
Comment (TomaszGaweda) [inline hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: Just to remind: We are adding here reflection cost

Comment (JackPGreen) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: > Just to remind: We are adding here reflection cost

I assumed that if were creating and discarding arrays here, that this wasn't a hot piece of code - but I've left the old method in place so we can switch this back if required.

**Comment by JamesHazelcast** [comment]
Hi @JackPGreen, thanks for putting this together, can I ask what led to these changes being proposed? While I understand that this change reduces the LoC needed to use the utility, it also introduces additional overhead that is entirely avoided by the extra couple of lines in the existing approach.

With Hazelcast being a high performance application, I'm not sure it's worth introducing overhead for such a minor code cleanup improvement. Sometimes we need to make some tidiness sacrifices in the pursuit of performance.

**Comment by JackPGreen** [comment]
> Hi @JackPGreen, thanks for putting this together, can I ask what led to these changes being proposed?

I wanted to combine two arrays in the UCD branch, but figured as this has some trivial value outside of that I'd spin it out on it's own instead.

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java (hunk 1 above)
Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: Please add a JMH benchmark.

And I'm not in favor of adding new methods if it isn't clear what kind of performance implication it has with respect to other methods.

Once the method is there, people will use it.

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: Since there currently is only 1 place where this is used, do we really want a new utility method at all?

Comment (pveentjer) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: > > Just to remind: We are adding here reflection cost
> 
> I _assumed_ that if were creating and discarding arrays here, that this wasn't a hot piece of code - but I've left the old method in place so we can switch this back if required.

You know what they say about assumption :) 

It is the mother of .... 

Also add this one on the JMH benchmark:

```
public static <T> T[] concat(T[] first, T... second) {
        T[] dst = (T[])new Object[first.length+second.length];
        concat(first, second, dst);
        return dst;
    }```

**Comment by JackPGreen** [comment]
> Please add a JMH benchmark.

```
package com.hazelcast.internal.metrics.impl;

import org.openjdk.jmh.annotations.*;
import com.hazelcast.internal.util.collection.ArrayUtils;
import java.lang.reflect.Array;
import java.util.concurrent.TimeUnit;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 200, time = 10, timeUnit = TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
public class ArrayUtilsBenchmark {
    private static String[] first = new String[] {"a", "b", "c"};
    private static String[] second = new String[] {"d", "e", "f"};

    @Benchmark
    public String[] baseCreate() {
        return createArrayNormally(first, second);
    }

    @Benchmark
    public String[] baseConcat() {
        String[] dest = baseCreate();
        ArrayUtils.concat(first, second, dest);
        return dest;
    }

    @Benchmark
    public String[] reflectCreate() {
        return createArrayReflectively(first, second);
    }

    @Benchmark
    public String[] reflectConcat() {
        String[] dest = reflectCreate();
        ArrayUtils.concat(first, second, dest);
        return dest;
    }

    private static String[] createArrayNormally(String[] first, String... second) {
        return new String[first.length + second.length];
    }

    private static <T> T[] createArrayReflectively(T[] first, T... second) {
        return (T[]) Array.newInstance(first.getClass().getComponentType(), first.length + second.length);
    }
}
```

Results:
| Benchmark  | Count | Time (ns/op) | gc.alloc.rate (MB/sec) |
|------------|-------|--------------|------------------------|
| Current    | 1000  | 29.89        | 1268                   |
| New Method | 1000  | 29.94        | 1267                   |

When you benchmark the array creation specifically:
| Benchmark           | Count | Time (ns/op) | gc.alloc.rate (MB/sec) |
|---------------------|-------|--------------|------------------------|
| `new String[]`      | 1000  | 3.086        | 12305                  |
| `Array.newInstance` | 1000  | 3.052        | 12438                  |

Basically the performance is identical.

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java (hunk 1 above)
Comment (JackPGreen) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: > Please add a JMH benchmark.

https://github.com/hazelcast/hazelcast/pull/25953#issuecomment-1805537416

 
> And I'm not in favor of adding new methods if it isn't clear what kind of performance implication it has with respect to other methods.
> 
> Once the method is there, people will use it.

Good point - I've used a similar method in `commons-lang3` and I'd never considered how it's able to do this.


> Since there currently is only 1 place where this is used, do we really want a new utility method at all?

It's used in OS (x1), EE (x1) and my UCD branch (x1).

> Also add this one on the JMH benchmark:
> 
> ```
> public static <T> T[] concat(T[] first, T... second) {
>         T[] dst = (T[])new Object[first.length+second.length];
>         concat(first, second, dst);
>         return dst;
>     }```
> ```

That's a good idea, but unfortunately doesn't work and throws a `ClassCastException`.

**Comment by JackPGreen** [comment]
This PR has been more divisive than expected - happy to abandon it.

File: hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java (hunk 1 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/query/impl/predicates/SqlPredicate.java:342]: Maybe:

```java
public static <T> T[] concat(T[] first, T... second, Supplier<T[]> supplier) {
        T[] dst = (T[]) supplier.get();
        concat(first, second, dst);
        return dst;
    }
```

Then call as 
`int[] abc = concat(a, b, int[]::new);
